client/sample-code/src/main/java/org/sagebionetworks/sample/sts/MigrateS3Bucket.java

```diff
@@ -0,0 +1,175 @@
+package org.sagebionetworks.sample.sts;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.sagebionetworks.client.SynapseClient;
+import org.sagebionetworks.client.exceptions.SynapseException;
+import org.sagebionetworks.repo.model.FileEntity;
+import org.sagebionetworks.repo.model.Folder;
+import org.sagebionetworks.repo.model.file.S3FileHandle;
+
+import java.io.IOException;
+import java.util.List;
+
+public class MigrateS3Bucket {
+	private static final String DELIMITER = "/";
+
+	private final DigestUtils md5DigestUtils;
+	private final AmazonS3 s3Client;
+	private final String s3Bucket;
+	private final String s3Folder;
+	private final long storageLocationId;
+	private final SynapseClient synapseClient;
+	private final String synapseFolderId;
+
+	public MigrateS3Bucket(AmazonS3 s3Client, SynapseClient synapseClient, String s3Bucket, String s3Folder,
+			String synapseFolderId, long storageLocationId) {
+		// Init MD5 digest utils.
+		this.md5DigestUtils = new DigestUtils(DigestUtils.getMd5Digest());
+
+		this.s3Client = s3Client;
+		this.synapseClient = synapseClient;
+
+		// This is the S3 bucket and folder that we migrate from. If folder is null, we import all files in the S3
+		// bucket.
+		this.s3Bucket = s3Bucket;
+		this.s3Folder = s3Folder;
+
+		// This is the entity ID of the folder in Synapse to which we migrate the S3 files. This folder must have an
+		// external S3 storage location with STS-enabled, as described in
+		// https://docs.synapse.org/articles/sts_storage_locations.html
+		this.synapseFolderId = synapseFolderId;
+
+		// This is the storage location ID that is set on the Synapse folder.
+		this.storageLocationId = storageLocationId;
+	}
+
+	/** Executes the migration. */
+	public void execute() throws IOException, SynapseException {
+		executeForSubfolder(s3Folder, synapseFolderId);
+	}
+
+	/**
+	 * Helper method to recursively walk the S3 folder hierarchy and migrate files to the given Synapse folder.
+	 * When this method makes a recursive call, it will update s3SubFolder to the next folder and it will create
+	 * a Synapse folder in which to migrate S3 files.
+	 *
+	 * @param s3SubFolder
+	 *         the S3 subfolder from which we migrate files; if this is null then we export from the S3 bucket root
+	 * @param synapseSubFolderId
+	 *         the Synapse folder to which we migrate the S3 files
+	 */
+	private void executeForSubfolder(String s3SubFolder, String synapseSubFolderId) throws IOException,
+			SynapseException {
+		ListObjectsRequest listObjectsRequest = new ListObjectsRequest().withBucketName(s3Bucket)
+				.withPrefix(s3SubFolder).withDelimiter(DELIMITER);
```

Comment:
Try prefix includes the trailing delimiter, so we don't need that weird if statement.

Reply:
I don't know what I did wrong when I first tested this, but apparently I don't need to filter out the subfolder at all.

Reply:
The documentation I found is inconclusive on this. On one side the official docs (https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ListObjectsRequest.html) seems to suggest that the delimiter is needed in the prefix to get our expected results but others suggest the behavior you reported initially. If I'm not wrong when you create a "folder" through the AWS console, AWS automatically adds a key named as the folder. The build is green so I'm guessing it's not an issue.